File: src/utils/riskAnalyzer.ts
export interface RiskData {
  returns: number[];
  prices: number[];
  market_returns: number[];
  portfolio_value: number;
}

export interface RiskMetrics {
  volatility: {
    daily_volatility: number;
    annual_volatility: number;
    downside_volatility: number;
    data_points: number;
  };
  var: {
    var_95_historical: number;
    var_99_historical: number;
    var_95_parametric: number;
    var_99_parametric: number;
    cvar_95: number;
    var_95_percent: number;
    var_99_percent: number;
  };
  drawdown: {
    max_drawdown: number;
    max_drawdown_percent: number;
    peak_price: number;
  };
  beta: {
    beta: number;
    alpha: number;
    r_squared: number;
    correlation: number;
  };
  sharpe: {
    daily_sharpe: number;
    annual_sharpe: number;
    risk_free_rate: number;
  };
  sortino: {
    daily_sortino: number;
    annual_sortino: number;
    downside_deviation: number;
  };
  stress_test: Record<string, { daily_return: string; portfolio_loss: number; loss_percent: number }>;
}

export class RiskAnalyzer {
  private warnings: string[] = [];
  private tradingDays = 252;

  calculateVolatility(returns: number[]): RiskMetrics['volatility'] {
    if (returns.length < 30) {
      this.warnings.push('收益率数据不足30天，波动率估计可能不准确');
    }

    const meanReturn = returns.reduce((a, b) => a + b, 0) / returns.length;
    const variance = returns.reduce((sum, r) => sum + Math.pow(r - meanReturn, 2), 0) / (returns.length - 1);
    const dailyVol = Math.sqrt(variance);
    const annualVol = dailyVol * Math.sqrt(this.tradingDays);

    const downsideReturns = returns.filter(r => r < 0);
    let downsideVol = 0;
    if (downsideReturns.length > 0) {
      const downsideMean = downsideReturns.reduce((a, b) => a + b, 0) / downsideReturns.length;
      const downsideVar = downsideReturns.reduce((sum, r) => sum + Math.pow(r - downsideMean, 2), 0) / downsideReturns.length;
      downsideVol = Math.sqrt(downsideVar) * Math.sqrt(this.tradingDays);
    }

    return {
      daily_volatility: Math.round(dailyVol * 10000) / 100,
      annual_volatility: Math.round(annualVol * 10000) / 100,
      downside_volatility: Math.round(downsideVol * 10000) / 100,
      data_points: returns.length
    };
  }

  calculateVaR(returns: number[], portfolioValue: number = 100000): RiskMetrics['var'] {
    if (returns.length < 100) {
      this.warnings.push('收益率数据不足100天，VaR估计可能不准确');
    }

    const sortedReturns = [...returns].sort((a, b) => a - b);
    const n = sortedReturns.length;

    const var95Index = Math.floor(n * 0.05);
    const var99Index = Math.floor(n * 0.01);

    const var95Return = sortedReturns[var95Index];
    const var99Return = sortedReturns[var99Index];

    const var95 = Math.abs(var95Return) * portfolioValue;
    const var99 = Math.abs(var99Return) * portfolioValue;

    const worstReturns = sortedReturns.slice(0, var95Index);
    const cvar95 = worstReturns.length > 0
      ? Math.abs(worstReturns.reduce((a, b) => a + b, 0) / worstReturns.length) * portfolioValue
      : 0;

    const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
    const std = Math.sqrt(returns.reduce((sum, r) => sum + Math.pow(r - mean, 2), 0) / (returns.length - 1));

    const var95Param = Math.abs(mean - 1.65 * std) * portfolioValue;
    const var99Param = Math.abs(mean - 2.33 * std) * portfolioValue;

    return {
      var_95_historical: Math.round(var95 * 100) / 100,
      var_99_historical: Math.round(var99 * 100) / 100,
      var_95_parametric: Math.round(var95Param * 100) / 100,
      var_99_parametric: Math.round(var99Param * 100) / 100,
      cvar_95: Math.round(cvar95 * 100) / 100,
      var_95_percent: Math.round(Math.abs(var95Return) * 10000) / 100,
      var_99_percent: Math.round(Math.abs(var99Return) * 10000) / 100
    };
  }

  calculateDrawdown(prices: number[]): RiskMetrics['drawdown'] {
    if (prices.length < 2) {
      return { max_drawdown: 0, max_drawdown_percent: 0, peak_price: prices[0] || 0 };
    }

    let peak = prices[0];
    let maxDrawdown = 0;
    let maxDrawdownPercent = 0;

    prices.forEach(price => {
      if (price > peak) {
        peak = price;
      }
      const drawdown = peak - price;
      const drawdownPercent = (drawdown / peak) * 100;

      if (drawdown > maxDrawdown) {
        maxDrawdown = drawdown;
        maxDrawdownPercent = drawdownPercent;
      }
    });

    return {
      max_drawdown: Math.round(maxDrawdown * 100) / 100,
      max_drawdown_percent: Math.round(maxDrawdownPercent * 100) / 100,
      peak_price: Math.round(peak * 100) / 100
    };
  }

  calculateBeta(stockReturns: number[], marketReturns: number[]): RiskMetrics['beta'] {
    if (stockReturns.length !== marketReturns.length) {
      throw new Error('股票收益率和市场收益率长度不一致');
    }

    const n = stockReturns.length;
    const stockMean = stockReturns.reduce((a, b) => a + b, 0) / n;
    const marketMean = marketReturns.reduce((a, b) => a + b, 0) / n;

    const covariance = stockReturns.reduce((sum, s, i) => {
      return sum + (s - stockMean) * (marketReturns[i] - marketMean);
    }, 0) / (n - 1);

    const marketVariance = marketReturns.reduce((sum, m) => {
      return sum + Math.pow(m - marketMean, 2);
    }, 0) / (n - 1);

    const beta = marketVariance !== 0 ? covariance / marketVariance : 0;
    const alpha = stockMean - beta * marketMean;

    const ssRes = stockReturns.reduce((sum, s, i) => {
      return sum + Math.pow(s - (alpha + beta * marketReturns[i]), 2);
    }, 0);

    const ssTot = stockReturns.reduce((sum, s) => {
      return sum + Math.pow(s - stockMean, 2);
    }, 0);

    const rSquared = ssTot !== 0 ? 1 - (ssRes / ssTot) : 0;

    return {
      beta: Math.round(beta * 100) / 100,
      alpha: Math.round(alpha * 10000) / 100,
      r_squared: Math.round(rSquared * 100) / 100,
      correlation: Math.round(Math.sqrt(rSquared) * (beta > 0 ? 1 : -1) * 100) / 100
    };
  }

  calculateSharpeRatio(returns: number[], riskFreeRate: number = 0.04): RiskMetrics['sharpe'] {
    const dailyRf = riskFreeRate / this.tradingDays;
    const meanReturn = returns.reduce((a, b) => a + b, 0) / returns.length;
    const excessReturn = meanReturn - dailyRf;

    const variance = returns.reduce((sum, r) => sum + Math.pow(r - meanReturn, 2), 0) / (returns.length - 1);
    const std = Math.sqrt(variance);

    const dailySharpe = std !== 0 ? excessReturn / std : 0;
    const annualSharpe = dailySharpe * Math.sqrt(this.tradingDays);

    return {
      daily_sharpe: Math.round(dailySharpe * 100) / 100,
      annual_sharpe: Math.round(annualSharpe * 100) / 100,
      risk_free_rate: riskFreeRate
    };
  }

  calculateSortinoRatio(returns: number[], riskFreeRate: number = 0.04): RiskMetrics['sortino'] {
    const dailyRf = riskFreeRate / this.tradingDays;
    const meanReturn = returns.reduce((a, b) => a + b, 0) / returns.length;
    const excessReturn = meanReturn - dailyRf;

    const downsideReturns = returns.filter(r => r < 0);
    let downsideStd = 0;
    if (downsideReturns.length > 0) {
      downsideStd = Math.sqrt(downsideReturns.reduce((sum, r) => sum + r * r, 0) / downsideReturns.length);
    }

    const dailySortino = downsideStd !== 0 ? excessReturn / downsideStd : 0;
    const annualSortino = dailySortino * Math.sqrt(this.tradingDays);

    return {
      daily_sortino: Math.round(dailySortino * 100) / 100,
      annual_sortino: Math.round(annualSortino * 100) / 100,
      downside_deviation: Math.round(downsideStd * Math.sqrt(this.tradingDays) * 10000) / 100
    };
  }

  stressTest(returns: number[], portfolioValue: number = 100000): RiskMetrics['stress_test'] {
    const scenarios: Record<string, number> = {
      '2008_financial_crisis': -0.20,
      '2020_covid_crash': -0.12,
      'flash_crash': -0.10,
      'moderate_decline': -0.05
    };

    const results: RiskMetrics['stress_test'] = {};

    Object.entries(scenarios).forEach(([scenario, dailyReturn]) => {
      const loss = Math.abs(dailyReturn) * portfolioValue;
      results[scenario] = {
        daily_return: `${(dailyReturn * 100).toFixed(0)}%`,
        portfolio_loss: Math.round(loss * 100) / 100,
        loss_percent: Math.round(Math.abs(dailyReturn) * 10000) / 100
      };
    });

    const sortedReturns = [...returns].sort((a, b) => a - b);
    const worst1Percent = sortedReturns.slice(0, Math.max(1, Math.floor(sortedReturns.length / 100)));
    const avgWorst1Percent = worst1Percent.length > 0
      ? worst1Percent.reduce((a, b) => a + b, 0) / worst1Percent.length
      : 0;

    results['historical_tail_1_percent'] = {
      daily_return: `${(avgWorst1Percent * 100).toFixed(2)}%`,
      portfolio_loss: Math.round(Math.abs(avgWorst1Percent) * portfolioValue * 100) / 100,
      loss_percent: Math.round(Math.abs(avgWorst1Percent) * 10000) / 100
    };

    return results;
  }

  comprehensiveRiskAnalysis(data: Partial<RiskData>): Partial<RiskMetrics> & { warnings: string[]; timestamp: string } {
    const returns = data.returns || [];
    const prices = data.prices || [];
    const marketReturns = data.market_returns || [];
    const portfolioValue = data.portfolio_value || 100000;

    if (returns.length === 0) {
      throw new Error('收益率数据不能为空');
    }

    this.warnings = [];

    const result: Partial<RiskMetrics> & { warnings: string[]; timestamp: string } = {
      volatility: this.calculateVolatility(returns),
      var: this.calculateVaR(returns, portfolioValue),
      sharpe: this.calculateSharpeRatio(returns),
      sortino: this.calculateSortinoRatio(returns),
      stress_test: this.stressTest(returns, portfolioValue),
      warnings: this.warnings,
      timestamp: new Date().toISOString()
    };

    if (prices.length > 0) {
      result.drawdown = this.calculateDrawdown(prices);
    }

    if (marketReturns.length > 0 && marketReturns.length === returns.length) {
      result.beta = this.calculateBeta(returns, marketReturns);
    }

    return result;
  }
}

export const riskAnalyzer = new RiskAnalyzer();

File: skills/investment-analyzer/scripts/risk_analyzer.py
#!/usr/bin/env python3
"""
Risk Analyzer - 风险分析模块

风险计算原则：
1. 多维度风险度量
2. 历史与隐含波动结合
3. 压力测试覆盖
4. 明确风险限额
"""

import json
import sys
import math
from typing import Dict, List, Optional, Tuple, Any
from dataclasses import dataclass
from datetime import datetime
import statistics


@dataclass
class RiskMetrics:
    """风险指标数据结构"""
    volatility: float = 0.0  # 年化波动率
    var_95: float = 0.0  # 95%置信度VaR
    var_99: float = 0.0  # 99%置信度VaR
    cvar_95: float = 0.0  # 95%CVaR
    max_drawdown: float = 0.0  # 最大回撤
    beta: float = 0.0  # Beta系数
    sharpe_ratio: float = 0.0  # 夏普比率
    sortino_ratio: float = 0.0  # Sortino比率


class RiskAnalyzer:
    """风险分析器"""

    def __init__(self):
        self.warnings = []
        self.trading_days = 252  # 年化交易日

    def calculate_volatility(self, returns: List[float]) -> Dict:
        """
        计算波动率

        Args:
            returns: 日收益率列表

        Returns:
            波动率指标
        """
        if len(returns) < 30:
            self.warnings.append("收益率数据不足30天，波动率估计可能不准确")

        # 计算日波动率
        mean_return = sum(returns) / len(returns)
        variance = sum((r - mean_return) ** 2 for r in returns) / (len(returns) - 1)
        daily_vol = math.sqrt(variance)

        # 年化波动率
        annual_vol = daily_vol * math.sqrt(self.trading_days)

        # 下行波动率
        downside_returns = [r for r in returns if r < 0]
        if downside_returns:
            downside_mean = sum(downside_returns) / len(downside_returns)
            downside_var = sum((r - downside_mean) ** 2 for r in downside_returns) / len(downside_returns)
            downside_vol = math.sqrt(downside_var) * math.sqrt(self.trading_days)
        else:
            downside_vol = 0

        return {
            "daily_volatility": round(daily_vol * 100, 2),
            "annual_volatility": round(annual_vol * 100, 2),
            "downside_volatility": round(downside_vol * 100, 2),
            "data_points": len(returns)
        }

    def calculate_var(self, returns: List[float], portfolio_value: float = 100000) -> Dict:
        """
        计算风险价值(VaR)

        Args:
            returns: 日收益率列表
            portfolio_value: 投资组合价值

        Returns:
            VaR指标
        """
        if len(returns) < 100:
            self.warnings.append("收益率数据不足100天，VaR估计可能不准确")

        sorted_returns = sorted(returns)
        n = len(sorted_returns)

        # 历史模拟法VaR
        var_95_index = int(n * 0.05)
        var_99_index = int(n * 0.01)

        var_95_return = sorted_returns[var_95_index]
        var_99_return = sorted_returns[var_99_index]

        # 转换为金额
        var_95 = abs(var_95_return) * portfolio_value
        var_99 = abs(var_99_return) * portfolio_value

        # CVaR (Expected Shortfall)
        cvar_95 = abs(sum(sorted_returns[:var_95_index]) / var_95_index) * portfolio_value if var_95_index > 0 else 0

        # 参数法VaR (正态分布假设)
        mean = sum(returns) / len(returns)
        std = math.sqrt(sum((r - mean) ** 2 for r in returns) / (len(returns) - 1))

        var_95_param = abs(mean - 1.65 * std) * portfolio_value
        var_99_param = abs(mean - 2.33 * std) * portfolio_value

        return {
            "var_95_historical": round(var_95, 2),
            "var_99_historical": round(var_99, 2),
            "var_95_parametric": round(var_95_param, 2),
            "var_99_parametric": round(var_99_param, 2),
            "cvar_95": round(cvar_95, 2),
            "var_95_percent": round(abs(var_95_return) * 100, 2),
            "var_99_percent": round(abs(var_99_return) * 100, 2),
            "portfolio_value": portfolio_value
        }

    def calculate_drawdown(self, prices: List[float]) -> Dict:
        """
        计算最大回撤

        Args:
            prices: 价格序列

        Returns:
            回撤指标
        """
        if len(prices) < 2:
            return {"max_drawdown": 0, "max_drawdown_percent": 0}

        peak = prices[0]
        max_drawdown = 0
        max_drawdown_percent = 0

        for price in prices:
            if price > peak:
                peak = price

            drawdown = peak - price
            drawdown_percent = (drawdown / peak) * 100

            if drawdown > max_drawdown:
                max_drawdown = drawdown
                max_drawdown_percent = drawdown_percent

        return {
            "max_drawdown": round(max_drawdown, 2),
            "max_drawdown_percent": round(max_drawdown_percent, 2),
            "peak_price": round(peak, 2)
        }

    def calculate_beta(self, stock_returns: List[float], market_returns: List[float]) -> Dict:
        """
        计算Beta系数

        Args:
            stock_returns: 股票收益率
            market_returns: 市场收益率

        Returns:
            Beta指标
        """
        if len(stock_returns) != len(market_returns):
            raise ValueError("股票收益率和市场收益率长度不一致")

        n = len(stock_returns)

        # 计算均值
        stock_mean = sum(stock_returns) / n
        market_mean = sum(market_returns) / n

        # 计算协方差和方差
        covariance = sum((s - stock_mean) * (m - market_mean) for s, m in zip(stock_returns, market_returns)) / (n - 1)
        market_variance = sum((m - market_mean) ** 2 for m in market_returns) / (n - 1)

        # Beta = Cov(r_stock, r_market) / Var(r_market)
        beta = covariance / market_variance if market_variance != 0 else 0

        # Alpha
        alpha = stock_mean - beta * market_mean

        # R-squared
        ss_res = sum((s - (alpha + beta * m)) ** 2 for s, m in zip(stock_returns, market_returns))
        ss_tot = sum((s - stock_mean) ** 2 for s in stock_returns)
        r_squared = 1 - (ss_res / ss_tot) if ss_tot != 0 else 0

        return {
            "beta": round(beta, 2),
            "alpha": round(alpha * 100, 2),
            "r_squared": round(r_squared, 2),
            "correlation": round(math.sqrt(r_squared) * (1 if beta > 0 else -1), 2)
        }

    def calculate_sharpe_ratio(self, returns: List[float], risk_free_rate: float = 0.04) -> Dict:
        """
        计算夏普比率

        Args:
            returns: 日收益率
            risk_free_rate: 年化无风险利率

        Returns:
            夏普比率
        """
        daily_rf = risk_free_rate / self.trading_days

        mean_return = sum(returns) / len(returns)
        excess_return = mean_return - daily_rf

        variance = sum((r - mean_return) ** 2 for r in returns) / (len(returns) - 1)
        std = math.sqrt(variance)

        daily_sharpe = excess_return / std if std != 0 else 0
        annual_sharpe = daily_sharpe * math.sqrt(self.trading_days)

        return {
            "daily_sharpe": round(daily_sharpe, 2),
            "annual_sharpe": round(annual_sharpe, 2),
            "risk_free_rate": risk_free_rate
        }

    def calculate_sortino_ratio(self, returns: List[float], risk_free_rate: float = 0.04) -> Dict:
        """
        计算Sortino比率

        Args:
            returns: 日收益率
            risk_free_rate: 年化无风险利率

        Returns:
            Sortino比率
        """
        daily_rf = risk_free_rate / self.trading_days

        mean_return = sum(returns) / len(returns)
        excess_return = mean_return - daily_rf

        # 下行标准差
        downside_returns = [r for r in returns if r < 0]
        if downside_returns:
            downside_std = math.sqrt(sum(r ** 2 for r in downside_returns) / len(downside_returns))
        else:
            downside_std = 0

        daily_sortino = excess_return / downside_std if downside_std != 0 else 0
        annual_sortino = daily_sortino * math.sqrt(self.trading_days)

        return {
            "daily_sortino": round(daily_sortino, 2),
            "annual_sortino": round(annual_sortino, 2),
            "downside_deviation": round(downside_std * math.sqrt(self.trading_days) * 100, 2)
        }

    def stress_test(self, returns: List[float], portfolio_value: float = 100000) -> Dict:
        """
        压力测试

        Args:
            returns: 日收益率
            portfolio_value: 投资组合价值

        Returns:
            压力测试结果
        """
        scenarios = {
            "2008_financial_crisis": -0.20,  # 单日下跌20%
            "2020_covid_crash": -0.12,  # 单日下跌12%
            "flash_crash": -0.10,  # 单日下跌10%
            "moderate_decline": -0.05,  # 单日下跌5%
        }

        results = {}
        for scenario, daily_return in scenarios.items():
            loss = abs(daily_return) * portfolio_value
            results[scenario] = {
                "daily_return": f"{daily_return*100:.0f}%",
                "portfolio_loss": round(loss, 2),
                "loss_percent": round(abs(daily_return) * 100, 2)
            }

        # 基于历史数据的尾部风险
        sorted_returns = sorted(returns)
        worst_1_percent = sorted_returns[:max(1, len(sorted_returns) // 100)]
        avg_worst_1_percent = sum(worst_1_percent) / len(worst_1_percent) if worst_1_percent else 0

        results["historical_tail_1_percent"] = {
            "avg_return": f"{avg_worst_1_percent*100:.2f}%",
            "portfolio_loss": round(abs(avg_worst_1_percent) * portfolio_value, 2)
        }

        return results

    def comprehensive_risk_analysis(self, data: Dict) -> Dict:
        """
        综合风险分析

        Args:
            data: {
                "returns": 日收益率列表,
                "prices": 价格序列,
                "market_returns": 市场收益率列表,
                "portfolio_value": 投资组合价值
            }

        Returns:
            综合风险分析结果
        """
        returns = data.get('returns', [])
        prices = data.get('prices', [])
        market_returns = data.get('market_returns', [])
        portfolio_value = data.get('portfolio_value', 100000)

        if not returns:
            raise ValueError("收益率数据不能为空")

        self.warnings = []

        result = {
            "volatility": self.calculate_volatility(returns),
            "var": self.calculate_var(returns, portfolio_value),
            "sharpe": self.calculate_sharpe_ratio(returns),
            "sortino": self.calculate_sortino_ratio(returns),
            "stress_test": self.stress_test(returns, portfolio_value),
            "warnings": self.warnings,
            "timestamp": datetime.now().isoformat()
        }

        if prices:
            result["drawdown"] = self.calculate_drawdown(prices)

        if market_returns and len(market_returns) == len(returns):
            result["beta"] = self.calculate_beta(returns, market_returns)

        return result


def main():
    """命令行入口"""
    if len(sys.argv) < 2:
        print("Usage: python risk_analyzer.py <command> [args...]")
        print("Commands:")
        print("  analyze <json_file>     - 综合风险分析")
        print("  volatility <json_file>  - 计算波动率")
        print("  var <json_file>         - 计算VaR")
        print("  stress <json_file>      - 压力测试")
        sys.exit(1)

    command = sys.argv[1]

    try:
        if len(sys.argv) < 3:
            print("Error: 需要JSON文件路径")
            sys.exit(1)

        with open(sys.argv[2], 'r') as f:
            data = json.load(f)

        analyzer = RiskAnalyzer()

        if command == "analyze":
            result = analyzer.comprehensive_risk_analysis(data)
        elif command == "volatility":
            returns = data.get('returns', [])
            result = analyzer.calculate_volatility(returns)
        elif command == "var":
            returns = data.get('returns', [])
            portfolio_value = data.get('portfolio_value', 100000)
            result = analyzer.calculate_var(returns, portfolio_value)
        elif command == "stress":
            returns = data.get('returns', [])
            portfolio_value = data.get('portfolio_value', 100000)
            result = analyzer.stress_test(returns, portfolio_value)
        else:
            print(f"Unknown command: {command}")
            sys.exit(1)

        print(json.dumps(result, indent=2, ensure_ascii=False))

    except Exception as e:
        print(json.dumps({"error": str(e)}, indent=2))
        sys.exit(1)


if __name__ == "__main__":
    main()
